pkg/daemon/config_drift_monitor.go

```diff
@@ -0,0 +1,378 @@
+		return nil
+	}
+
+	if err := validateOnDiskState(c.MachineConfig, c.SystemdPath); err != nil {
+		return configDriftErr(err)
+	}
+
+	return nil
+}
+
+// Finds the paths for all files in a given MachineConfig.
+func getFilePathsFromMachineConfig(mc *mcfgv1.MachineConfig, systemdPath string) (sets.String, error) {
+	ignConfig, err := ctrlcommon.IgnParseWrapper(mc.Spec.Config.Raw)
+	if err != nil {
+		return sets.String{}, fmt.Errorf("could not get dirs from ignition config: %w", err)
+	}
+
+	switch typedConfig := ignConfig.(type) {
+	case ign3types.Config:
+		return getFilePathsFromIgn3Config(ignConfig.(ign3types.Config), systemdPath), nil
+	case ign2types.Config:
+		return getFilePathsFromIgn2Config(ignConfig.(ign2types.Config), systemdPath), nil
+	default:
+		return sets.String{}, fmt.Errorf("unexpected type for ignition config: %v", typedConfig)
+	}
+}
+
+// Extracts all unique directories from a given Ignition 3 config.
+func getFilePathsFromIgn3Config(ignConfig ign3types.Config, systemdPath string) sets.String {
+	files := sets.NewString()
+
+	// Get all the file paths from the ignition config
+	for _, ignFile := range ignConfig.Storage.Files {
+		if _, err := os.Stat(ignFile.Path); err == nil && !os.IsNotExist(err) {
+			files.Insert(ignFile.Path)
+		}
+	}
+
+	// Get all the file paths for systemd dropins from the ignition config
+	for _, unit := range ignConfig.Systemd.Units {
+		for _, dropin := range unit.Dropins {
+			dropinPath := getIgn3SystemdPath(systemdPath, unit, dropin)
+			if _, err := os.Stat(dropinPath); err == nil && !os.IsNotExist(err) {
+				files.Insert(dropinPath)
+			}
+		}
+	}
+
+	return files
+}
+
+// Extracts all unique directories from a given Ignition 2 config.
+func getFilePathsFromIgn2Config(ignConfig ign2types.Config, systemdPath string) sets.String {
+	files := sets.NewString()
+
+	// Get all the file paths from the ignition config
+	for _, ignFile := range ignConfig.Storage.Files {
+		if _, err := os.Stat(ignFile.Path); err == nil && !os.IsNotExist(err) {
+			files.Insert(ignFile.Path)
+		}
+	}
+
+	// Get all the file paths for systemd dropins from the ignition config
+	for _, unit := range ignConfig.Systemd.Units {
+		for _, dropin := range unit.Dropins {
+			dropinPath := getIgn2SystemdPath(systemdPath, unit, dropin)
+			if _, err := os.Stat(dropinPath); err == nil && !os.IsNotExist(err) {
+				files.Insert(dropinPath)
+			}
+		}
+	}
+
+	return files
+}
+
+// Gets the directories for all the MachineConfig file paths while
+// deduplicating them.
+func getDirPathsFromFilePaths(filePaths sets.String) []string {
+	dirPaths := sets.NewString()
+
+	for filePath := range filePaths {
+		dirPaths.Insert(filepath.Dir(filePath))
+	}
+
+	return dirPaths.List()
+}
+
+// Computes the absolute path for a given systemd unit and dropin, given an
+// optional root path which defaults to /etc/systemd.
+func getIgn2SystemdPath(systemdPath string, unit ign2types.Unit, dropin ign2types.SystemdDropin) string {
```

Comment:
are these funcs only called but the getfilepaths? if so I think we could just combine them into those funcs since the code is trivial.

Reply:
No, I use these in my tests. I was also planning to add them to `checkV2Units` and `checkV3Units`, respectively.

My intent is to avoid duplicating the `filepath.Join` calls in multiple places especially since we manipulate one of the values passed into `filepath.Join` (ign3types.Systemd.Name + ".d").

Reply:
i think i missed that, will relook :+1: